exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp

```diff
@@ -0,0 +1,230 @@
+{
+    const int T = prob_->get_T();
+    const int NX = prob_->get_num_positions();
+    const int NU = prob_->get_num_controls();
+    const int N = NX + prob_->get_num_velocities();
+
+    prob_->ResetCostEvolution(GetNumberOfMaxIterations() + 1);
+    if (!prob_) ThrowNamed("Solver has not been initialized!");
+
+    // initialize Gain matrices
+    K_gains_.assign(T, Eigen::MatrixXd(NU, N));
+    Ku_gains_.assign(T, Eigen::MatrixXd(NU, NU));
+    Kv_gains_.assign(T, Eigen::MatrixXd(NU, N));
+    vk_gains_.assign(T, Eigen::MatrixXd(N, 1));
+
+    // all of the below are not pointers, since we want to copy over
+    //  solutions across iterations
+    Eigen::MatrixXd best_U, best_X, global_best_U;
+    solution.resize(T, NU);
+
+    if (debug_) HIGHLIGHT_NAMED("ILQRSolver", "Running ILQR solver for max " << parameters_.MaxIterations << " iterations");
+
+    double last_cost = 0, global_best_cost = 0;
+    int last_best_iteration = 0;
+
+    for (int iteration = 0; iteration < GetNumberOfMaxIterations(); ++iteration)
+    {
+        // Backwards pass computes the gains
+        BackwardPass();
+        if (debug_) HIGHLIGHT_NAMED("ILQRSolver", "Backward pass complete");
+
+        // forward pass to compute new control trajectory
+        // TODO: Configure line-search space from xml
+        // TODO (Wolf): What you are doing is a forward line-search when we may try a
+        //    backtracking line search for a performance improvement later - this just as an aside.
+        const Eigen::VectorXd alpha_space = Eigen::VectorXd::LinSpaced(10, 0.1, 1.0);
+        double best_cost = 0, best_alpha = 0;
+
+        Eigen::MatrixXd ref_trajectory = prob_->get_X();
+        // perform a linear search to find the best rate
+        for (int ai = 0; ai < alpha_space.rows(); ++ai)
+        {
+            double alpha = alpha_space(ai);
+            double cost = ForwardPass(alpha, ref_trajectory);
+
+            if (ai == 0 || cost < best_cost)
+            {
+                best_cost = cost;
+                best_X = prob_->get_X();
+                best_U = prob_->get_U();
+                best_alpha = alpha;
+            }
+        }
+
+        if (debug_)
+        {
+            HIGHLIGHT_NAMED("ILQRSolver", "Forward pass complete with cost: " << best_cost << " and alpha " << best_alpha);
+            HIGHLIGHT_NAMED("ILQRSolver", "Final state: " << best_X.col(T - 1).transpose());
+        }
+
+        // copy solutions for next iteration
+        if (iteration == 0 || global_best_cost > best_cost)
+        {
+            global_best_cost = best_cost;
+            last_best_iteration = iteration;
+            global_best_U = best_U;
+        }
+
+        if (iteration - last_best_iteration > parameters_.FunctionTolerancePatience)
+        {
+            if (debug_) HIGHLIGHT_NAMED("ILQRSolver", "Early stopping criterion reached.");
+            break;
+        }
+
+        if (abs(best_cost - last_cost) < parameters_.FunctionTolerance)
```

Comment:
`std::abs`, and check the other solvers (AICO), we want the relative not the absolute tolerance

Reply:
I changed the logic a bit here. I'm allowing the cost to increase in case of local minima. This is different from AICO:
```
(f_t-1 - f_t) <= functionTolerance * max(1, abs(f_t))
```
where a negative cost will always result in terimnation.

Reply:
Correct. I did the same in snopt but confused it with aico. Good job!